### ISSUE TITLE
A2C seems to be incapable of learning Cartpole

### ISSUE THREAD

**Issue by stheid** [tensorforce/tensorforce#702]
Maybe this is a stupid issue, but i have the impression that there is something wrong.

To my knowledge PPO can be seen as an extension to A2C and additionally constraints the learning task. Therefore i would expect A2C to learn at least comparable well to PPO if not clearly outperform it.

The appendet script shows no real learning after 100 or 200 episodes, while changing the `agent` parameter to PPO will nearly solve the task in 100 episodes.

![image](https://user-images.githubusercontent.com/2736207/84759714-63722880-afc7-11ea-9c33-da45ca6e268d.png)
orange is ppo, blue is a2c

additionally some plots show an offset by 10 episodes (x axis) in the data, even though both recordings are carried out with the exact script below with only the agent changed.
![image](https://user-images.githubusercontent.com/2736207/84759804-8270ba80-afc7-11ea-962a-789dbb926af7.png)

```python
if __name__ == '__main__':
    environment = Environment.create(environment='gym', level='CartPole-v1')
    agent = Agent.create(
        agent='a2c', environment=environment, batch_size=10, learning_rate=1e-3,summarizer=dict(
        directory='data/summaries',
        # list of labels, or 'all'
        labels=['graph', 'entropy', 'kl-divergence', 'losses', 'rewards'],
        frequency=100,  # store values every 100 timesteps
    ))

    # Initialize the runner
    runner = Runner(agent=agent, environment=environment)

    # Start the runner
    runner.run(num_episodes=100)
    environment.visualize = True
    runner.run(num_episodes=1, evaluation=True)
    runner.close()
```

let me know if this makes any sense...

**Comment by AlexKuhnle** [comment]
A2C and PPO are not really the same, or at least they are not in Tensorforce (I think AC itself is a somewhat vague term, so Tensorforce's PPO is probably also AC according to some interpretation). The key difference is that PPO follows the REINFORCE-style of using the empirical rollout return ("reward to go") as return, and uses the baseline as, well, baseline if specified. AC and A2C both don't use the empirical return, but the predicted state/Q-value per timestep (and A2C additionally subtracts the critic estimate as baseline value, similar to PPO). AC and A2C hence do timestep-based updates (since they don't need to wait for a full episode to compute returns) and use a replay memory (typical for "DQN-style" value functions, since they can make use of off-policy data). One practical aspect to keep in mind: batch_size and update_frequency for A(2)C are in timesteps, whereas they are in episodes for PPO/VPG/TRPO.

If there's any part where more detail would be helpful, let me know.

**Comment by stheid** [comment]
Thanks for the explaination, i think i dont fully understand, but thats due to the fact that i am still not very experienced in the field.

However, it really confuses me that A2C is incapable of solving Cartpole, while it works fine with https://github.com/hill-a/stable-baselines A2C implementation. But again you are probably right that there is also a lot of configuration and fine implementation decisions that can yield to different results in different libraries.

**Comment by stheid** [comment]
Maybe one more question. What algorithm of your library would you recommend as a starting point for a task with continous actions?

**Comment by twmmason** [comment]
unsubscribe

**Comment by AlexKuhnle** [comment]
PPO is always a good choice, and DDPG is generally also supposed to be good for continuous actions. Regarding your config above, the `batch_size` of 10 is definitely quite small (as I mentioned before, it's timesteps here whereas episodes for PPO). Also, the default parameters are probably not great, and in general hyperparameters should be tuned to the problem. I would expect that this implementation of A2C is also able in principle to solve CartPole.

Baselines with baselines zoo have spent quite some time on identifying good hyperparameters for some standard benchmarks, which hasn't happened recently as much with Tensorforce (mainly because I don't have enough time -- involvement on this front would be very welcome :-). I plan to focus more on that with 0.6, when the framework is finally properly compatible with TF2. But again, any involvement and contribution (configs, example scripts) here is very welcome, and I would be happy to work together on this.

**Comment by twmmason** [comment]
Unsubscribe

**Comment by ppxyn1** [comment]
@stheid Hi, I am struggling with the same issue with A2C in Cart-pole env.
Results are different from experiments in different libraries with the same settings. 
In my code, results are not optimized with Tensorforce`s A2C.
I wonder if you managed to implement A2C to solve CartPole.
